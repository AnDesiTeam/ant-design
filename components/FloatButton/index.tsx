import VerticalAlignTopOutlined from '@ant-design/icons/VerticalAlignTopOutlined';
import classNames from 'classnames';
import CSSMotion from 'rc-motion';
import addEventListener from 'rc-util/lib/Dom/addEventListener';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import omit from 'rc-util/lib/omit';
import * as React from 'react';
import { ConfigContext } from '../config-provider';
import getScroll from '../_util/getScroll';
import { cloneElement } from '../_util/reactNode';
import scrollTo from '../_util/scrollTo';
import { throttleByAnimationFrame } from '../_util/throttleByAnimationFrame';
import useStyle from './style';

export interface FloatButtonProps {
  visibilityHeight?: number;
  onClick?: React.MouseEventHandler<HTMLElement>;
  target?: () => HTMLElement | Window | Document;
  prefixCls?: string;
  children?: React.ReactNode;
  className?: string;
  style?: React.CSSProperties;
  duration?: number;
  visible?: boolean; // Only for test. Don't use it.
}

interface ContentProps {
  prefixCls: string;
  rootPrefixCls: string;
  children?: React.ReactNode;
  visible?: boolean; // Only for test. Don't use it.
}

const FloatButtonContent: React.FC<ContentProps> = props => {
  const { prefixCls, rootPrefixCls, children, visible } = props;
  const defaultElement = (
    <div className={`${prefixCls}-content`}>
      <div className={`${prefixCls}-icon`}>
        <VerticalAlignTopOutlined />
      </div>
    </div>
  );
  return (
    <CSSMotion visible={visible} motionName={`${rootPrefixCls}-fade`}>
      {({ className: motionClassName }) =>
        cloneElement(children || defaultElement, ({ className }) => ({
          className: classNames(motionClassName, className),
        }))
      }
    </CSSMotion>
  );
};

const FloatButton: React.FC<FloatButtonProps> = props => {
  const [visible, setVisible] = useMergedState(false, {
    value: props.visible,
  });

  const ref = React.createRef<HTMLDivElement>();
  const scrollEvent = React.useRef<any>();

  const getDefaultTarget = () =>
    ref.current && ref.current.ownerDocument ? ref.current.ownerDocument : window;

  const handleScroll = throttleByAnimationFrame(
    (e: React.UIEvent<HTMLElement> | { target: any }) => {
      const { visibilityHeight } = props;
      const scrollTop = getScroll(e.target, true);
      setVisible(scrollTop > visibilityHeight!);
    },
  );

  const bindScrollEvent = () => {
    const { target } = props;
    const getTarget = target || getDefaultTarget;
    const container = getTarget();
    scrollEvent.current = addEventListener(container, 'scroll', (e: React.UIEvent<HTMLElement>) => {
      handleScroll(e);
    });
    handleScroll({ target: container });
  };

  React.useEffect(() => {
    bindScrollEvent();
    return () => {
      if (scrollEvent.current) {
        scrollEvent.current.remove();
      }
      handleScroll.cancel();
    };
  }, [props.target]);

  const scrollToTop = (e: React.MouseEvent<HTMLDivElement>) => {
    const { onClick, target, duration = 450 } = props;
    scrollTo(0, {
      getContainer: target || getDefaultTarget,
      duration,
    });
    if (typeof onClick === 'function') {
      onClick(e);
    }
  };

  const { getPrefixCls, direction } = React.useContext(ConfigContext);
  const { prefixCls: customizePrefixCls, className = '' } = props;
  const prefixCls = getPrefixCls('back-top', customizePrefixCls);
  const rootPrefixCls = getPrefixCls();
  const [wrapSSR, hashId] = useStyle(prefixCls);

  const classString = classNames(hashId, prefixCls, className, {
    [`${prefixCls}-rtl`]: direction === 'rtl',
  });

  // fix https://fb.me/react-unknown-prop
  const divProps = omit(props, [
    'prefixCls',
    'className',
    'children',
    'visibilityHeight',
    'target',
    'visible',
  ]);

  return wrapSSR(
    <div {...divProps} className={classString} onClick={scrollToTop} ref={ref}>
      <FloatButtonContent prefixCls={prefixCls} rootPrefixCls={rootPrefixCls} visible={visible}>
        {props.children}
      </FloatButtonContent>
    </div>,
  );
};

FloatButton.defaultProps = {
  visibilityHeight: 400,
};

export default React.memo(FloatButton);
