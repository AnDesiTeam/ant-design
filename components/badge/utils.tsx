import { PresetColorTypes } from '../_util/colors';
import type { PresetColorType } from '../_util/colors';

// eslint-disable-next-line import/prefer-default-export
export function isPresetColor(color?: string): boolean {
  return PresetColorTypes.includes(color as PresetColorType);
}
