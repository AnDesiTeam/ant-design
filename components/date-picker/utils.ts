import * as moment from 'moment';

export function formatDate(
  value: moment.Moment | undefined | null,
  format: string | string[] | ((moment: moment.Moment | undefined) => string),
): string {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }

  if (typeof format === 'function') {
    format = format(value);
  }

  return value.format(format);
}
