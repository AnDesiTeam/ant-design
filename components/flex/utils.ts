import classNames from 'classnames';

import type { FlexProps } from './interface';

export const flexDirectionValues = ['row', 'row-reverse', 'column', 'column-reverse'] as const;

export const flexWrapValues = ['wrap', 'nowrap', 'wrap-reverse'] as const;

export const justifyContentValues = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
] as const;

export const alignItemsValues = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
] as const;

const genClsWrap = (prefixCls: string, wrap: FlexProps['wrap']) => {
  const wrapCls: Record<PropertyKey, boolean> = {};
  flexWrapValues.forEach((cssKey) => {
    wrapCls[`${prefixCls}-wrap-${cssKey}`] = wrap === cssKey;
  });
  return wrapCls;
};

const genClsAlign = (prefixCls: string, align: FlexProps['align']) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach((cssKey) => {
    alignCls[`${prefixCls}-align-${cssKey}`] = align === cssKey;
  });
  return alignCls;
};

const genClsJustify = (prefixCls: string, justify: FlexProps['justify']) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach((cssKey) => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = justify === cssKey;
  });
  return justifyCls;
};

const genClsDirection = (prefixCls: string, direction: FlexProps['direction']) => {
  const directionCls: Record<PropertyKey, boolean> = {};
  flexDirectionValues.forEach((cssKey) => {
    directionCls[`${prefixCls}-direction-${cssKey}`] = direction === cssKey;
  });
  return directionCls;
};

const genMergedAlignCls = (
  prefixCls: string,
  align: FlexProps['align'],
  direction: FlexProps['direction'],
) => {
  if (align) {
    return genClsAlign(prefixCls, align);
  }
  return {
    ...genClsAlign(prefixCls, align),
    [`${prefixCls}-align-stretch`]: direction === 'column' || direction === 'column-reverse',
  };
};

const createFlexClassNames = (
  prefixCls: string,
  props: Pick<FlexProps, 'direction' | 'wrap' | 'justify' | 'align'>,
) => {
  const { direction, wrap, justify, align } = props;
  return classNames({
    ...genClsWrap(prefixCls, wrap),
    ...genClsJustify(prefixCls, justify),
    ...genClsDirection(prefixCls, direction),
    ...genMergedAlignCls(prefixCls, align, direction),
  });
};

export default createFlexClassNames;
