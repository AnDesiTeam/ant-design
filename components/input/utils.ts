import type { ReactNode } from 'react';
import type { InputProps } from './Input';

// eslint-disable-next-line import/prefer-default-export
export function hasPrefixSuffix(props: {
  prefix?: ReactNode;
  suffix?: ReactNode;
  allowClear?: InputProps['allowClear'];
}) {
  return !!(props.prefix || props.suffix || props.allowClear);
}
