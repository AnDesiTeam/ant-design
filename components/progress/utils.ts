import { presetPrimaryColors } from '@ant-design/colors';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';
import warning from '../_util/warning';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    warning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({ success = {}, strokeColor }: Partial<CircleProps>): (
  | string
  | Record<PropertyKey, string>
)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
  },
): [number, number] => {
  let width: number = -1;
  let height: number = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = size === 'small' ? 6 : 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return [width, height];
};
