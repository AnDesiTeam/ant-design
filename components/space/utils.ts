import type { SpaceSize } from '.';
import type { SizeType } from '../config-provider/SizeContext';

export const isPresetSize = (size: SpaceSize): size is SizeType => {
  if (!size) {
    return false;
  }
  return typeof size === 'string' && ['small', 'middle', 'large'].includes(size);
};

export const isValidNumber = (size: SpaceSize): size is number => {
  if (!size) {
    // 这里故意排除 size = 0 的情况，因为 gap 属性在 CSS 中的默认值就是 0，所以如果用户传0进来的话，我们可以直接忽略
    return false;
  }
  return typeof size === 'number' && !Number.isNaN(size);
};
