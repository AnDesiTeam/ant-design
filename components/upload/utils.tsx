import { RcFile, UploadFile } from './interface';

export function T() {
  return true;
}

// Fix IE file.status problem
// via coping a new Object
export function fileToObject(file: RcFile): UploadFile {
  return {
    ...file,
    lastModified: file.lastModified,
    lastModifiedDate: file.lastModifiedDate,
    name: file.name,
    size: file.size,
    type: file.type,
    uid: file.uid,
    percent: 0,
    originFileObj: file,
  } as UploadFile;
}

/**
 * 生成Progress percent: 0.1 -> 0.98
 *   - for ie
 */
export function genPercentAdd() {
  let k = 0.1;
  const i = 0.01;
  const end = 0.98;
  return function(s: number) {
    let start = s;
    if (start >= end) {
      return start;
    }

    start += k;
    k = k - i;
    if (k < 0.001) {
      k = 0.001;
    }
    return start;
  };
}

export function getFileItem(file: UploadFile, fileList: UploadFile[]) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  return fileList.filter(item => item[matchKey] === file[matchKey])[0];
}

export function removeFileItem(file: UploadFile, fileList: UploadFile[]) {
  const matchKey = file.uid !== undefined ? 'uid' : 'name';
  const removed = fileList.filter(item => item[matchKey] !== file[matchKey]);
  if (removed.length === fileList.length) {
    return null;
  }
  return removed;
}

// ===================== Process Image Preview =====================
const BUFFER_CHUNK_SIZE = 1024 * 500;
const BASE64_CHUNK_SIZE = 3 * 1024 * 200; // Must be 3x

function readBuffers(file: File | Blob): Promise<any> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    let offset = 0;
    let binary: string = '';

    reader.onload = function() {
      const buffer = new Uint8Array(reader.result as ArrayBuffer);
      const len = buffer.byteLength;
      for (let i = 0; i < len; i++) {
        binary += String.fromCharCode(buffer[i]);
      }
      offset += BUFFER_CHUNK_SIZE;
      seek();
    };
    reader.onerror = function(event) {
      reader.abort();
      reject(event);
    };
    seek();

    function seek() {
      if (offset >= file.size) {
        resolve(binary);
        return;
      }
      const slice = file.slice(offset, offset + BUFFER_CHUNK_SIZE);
      reader.readAsArrayBuffer(slice);
    }
  });
}

/**
 * Async convert file to base64 image.
 * We split file into chunk which is size of 3x to make base64 work.
 */
export function previewImage(file: File | Blob): PromiseLike<any> {
  return readBuffers(file).then(
    (binary: string) =>
      new Promise(resolve => {
        let baseStr: string = '';
        let offset: number = 0;

        // Use macro task here to prevent it block UI render.
        // TODO: Optimize this with `requestIdleCallback` after we don't support old browser.
        let id: number;
        function step() {
          const subStr = binary.substr(offset, BASE64_CHUNK_SIZE);
          baseStr += btoa(subStr);
          offset += BASE64_CHUNK_SIZE;

          if (offset > binary.length) {
            window.clearInterval(id);
            resolve(`data:image;base64, ${baseStr}`);
          }
        }

        id = window.setInterval(step, 5);
        step();
      }),
  );
}
