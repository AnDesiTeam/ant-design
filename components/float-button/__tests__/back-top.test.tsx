import React from 'react';
import FloatButton from '..';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { fireEvent, render } from '../../../tests/utils';

const { BackTop } = FloatButton;
describe('BackTop', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });
  mountTest(BackTop);
  rtlTest(BackTop);

  it('should scroll to top after click it', () => {
    const { container } = render(<BackTop visibilityHeight={-1} />);
    const scrollToSpy = jest.spyOn(window, 'scrollTo').mockImplementation((_, y) => {
      window.scrollY = y;
      window.pageYOffset = y;
      document.documentElement.scrollTop = y;
    });
    window.scrollTo(0, 400);
    jest.runAllTimers();
    expect(document.documentElement.scrollTop).toBe(400);
    fireEvent.click(container.querySelector<HTMLButtonElement>('.ant-float-btn')!);
    jest.runAllTimers();
    expect(document.documentElement.scrollTop).toBe(0);
    scrollToSpy.mockRestore();
  });

  it('support onClick', () => {
    const onClick = jest.fn();
    const { container } = render(<BackTop visibilityHeight={0} onClick={onClick} />);
    fireEvent.click(container.querySelector<HTMLButtonElement>('.ant-float-btn')!);
    expect(onClick).toHaveBeenCalled();
  });

  it('support invalid target', () => {
    const onClick = jest.fn();
    const { container } = render(
      <BackTop onClick={onClick} visibilityHeight={0} target={undefined} />,
    );
    fireEvent.click(container.querySelector<HTMLButtonElement>('.ant-float-btn')!);
    expect(onClick).toHaveBeenCalled();
  });

  it('pass style to float button', () => {
    const { container } = render(<BackTop style={{ color: 'red' }} visibilityHeight={0} />);
    expect(container.querySelector<HTMLButtonElement>('.ant-float-btn')?.style.color).toBe('red');
  });
});
