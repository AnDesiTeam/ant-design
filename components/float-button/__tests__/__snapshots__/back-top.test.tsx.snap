// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackTop rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-float-btn ant-float-btn-default ant-float-btn-circle ant-float-btn-rtl"
>
  <div
    class="ant-float-btn-body"
  >
    <div
      class="ant-float-btn-content"
    >
      <div
        class="ant-float-btn-content"
      >
        <div
          class="ant-float-btn-icon ant-float-btn-default-icon"
        >
          <span
            aria-label="vertical-align-top"
            class="anticon anticon-vertical-align-top"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="vertical-align-top"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M859.9 168H164.1c-4.5 0-8.1 3.6-8.1 8v60c0 4.4 3.6 8 8.1 8h695.8c4.5 0 8.1-3.6 8.1-8v-60c0-4.4-3.6-8-8.1-8zM518.3 355a8 8 0 00-12.6 0l-112 141.7a7.98 7.98 0 006.3 12.9h73.9V848c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V509.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 355z"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
